From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 22 Sep 2020 01:49:19 -0700
Subject: [PATCH] Optimise non-flush packet sending

Ported from Winds-Studio/Leaf ver/26.2 (0202-Optimise-non-flush-packet-sending.patch), GPL-3.0-only.
Original author: Spottedleaf.
Canvas adaptation: config gate mapped to optimizations.nonFlushPacketSending.

diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -128,6 +128,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public boolean isPending = true;
     public boolean queueImmunity;
     // Paper end - Optimize network
+    private io.netty.channel.SingleThreadEventLoop eventLoop; // Paper - optimise packets that are not flushed
 
     public Connection(final PacketFlow receiving) {
         this.receiving = receiving;
@@ -137,6 +138,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public void channelActive(final ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         this.channel = ctx.channel();
+        this.eventLoop = (io.netty.channel.SingleThreadEventLoop) this.channel.eventLoop(); // Paper - optimise packets that are not flushed
         this.address = this.channel.remoteAddress();
         this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
@@ -487,6 +489,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, listener, flush);
         } else {
+            // Paper start - optimise packets that are not flushed
+            if (!flush && io.canvasmc.canvas.GlobalConfiguration.getInstance().optimizations.nonFlushPacketSending) { // Canvas - optimise packets that are not flushed
+                this.eventLoop.lazyExecute(() -> this.doSendPacket(packet, listener, false));
+            } else
+                // Paper end - optimise packets that are not flushed
             this.channel.eventLoop().execute(() -> this.doSendPacket(packet, listener, flush));
         }
     }
